src/raw.rs

```diff
@@ -27,415 +311,546 @@ where
-        self.reverse = true;
+    pub fn key_only(mut self) -> Self {
+        self.state.inner_mut().map(|mut x| {
+            x.key_only = true;
+            x
+        });
         self
     }
 }
 
-impl<'a> Future for Scan<'a> {
+impl Future for Scan {
     type Item = Vec<KvPair>;
-    type Error = ();
+    type Error = Error;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        let _ = &self.client;
-        let _ = &self.range;
-        let _ = &self.limit;
-        let _ = &self.key_only;
-        let _ = &self.cf;
-        unimplemented!()
+        self.state.poll()
     }
 }
 
-pub struct BatchScan<'a> {
-    client: &'a Client,
-    ranges: Vec<(Key, Key)>,
+pub(crate) struct BatchScanInner {
+    ranges: Vec<Result<(Key, Option<Key>)>>,
     each_limit: u32,
     key_only: bool,
-    cf: Option<ColumnFamily>,
-    reverse: bool,
 }
 
-impl<'a> BatchScan<'a> {
-    fn new(client: &'a Client, ranges: Vec<(Key, Key)>, each_limit: u32) -> Self {
-        BatchScan {
-            client,
+impl BatchScanInner {
+    fn new(ranges: Vec<Result<(Key, Option<Key>)>>, each_limit: u32) -> Self {
+        BatchScanInner {
             ranges,
             each_limit,
             key_only: false,
-            cf: None,
-            reverse: false,
         }
     }
+}
 
-    pub fn key_only(mut self) -> Self {
-        self.key_only = true;
-        self
+impl RequestInner for BatchScanInner {
+    type Resp = Vec<KvPair>;
+
+    fn execute(self, client: Arc<RpcClient>, cf: Option<ColumnFamily>) -> KvFuture<Self::Resp> {
+        let (errors, ranges): (Vec<_>, Vec<_>) = self.ranges.into_iter().partition(Result::is_err);
+        if !errors.is_empty() {
+            // All errors must be InvalidKeyRange so we can simply return a new InvalidKeyRange
+            Box::new(future::err(Error::InvalidKeyRange))
+        } else {
+            Box::new(client.raw_batch_scan(
+                ranges.into_iter().map(Result::unwrap).collect(),
+                self.each_limit,
+                self.key_only,
+                cf,
+            ))
+        }
+    }
+}
+
+/// An unresolved [`Client::batch_scan`](struct.Client.html#method.batch_scan) request.
+///
+/// Once resolved this request will result in a scanner over the given ranges.
+pub struct BatchScan {
+    state: RequestState<BatchScanInner>,
+}
+
+impl BatchScan {
+    fn new(client: Arc<RpcClient>, inner: BatchScanInner) -> Self {
+        BatchScan {
+            state: RequestState::new(client, inner),
+        }
     }
 
+    /// Set the (optional) [`ColumnFamily`](struct.ColumnFamily.html).
     pub fn cf(mut self, cf: impl Into<ColumnFamily>) -> Self {
-        self.cf = Some(cf.into());
+        self.state.cf(cf);
         self
     }
 
-    pub fn reverse(mut self) -> Self {
-        self.reverse = true;
+    pub fn key_only(mut self) -> Self {
+        self.state.inner_mut().map(|mut x| {
+            x.key_only = true;
+            x
+        });
```

Comment:
Another use of map that could be simplified.

Reply:
Fixed